src/headerssync.h

```diff
@@ -0,0 +1,255 @@
+ *
+ * - In phase 2 (redownload), keep a lookahead buffer and only accept headers
+ * from that buffer into the block index (permanent memory usage) once they
+ * have some target number of verified commitments on top of them. With this
+ * parametrization, we can achieve a given security target for potential
+ * permanent memory usage, while choosing N to minimize memory use during the
+ * sync (temporary, per-peer storage).
+ */
+
+class HeadersSyncState {
+public:
+    HeadersSyncState(NodeId id, const Consensus::Params& consensus_params);
+    ~HeadersSyncState() {}
+
+    enum class State {
+        UNSTARTED,
+        /** INITIAL_DOWNLOAD means the peer has not yet demonstrated their
+         * chain has sufficient work */
+        INITIAL_DOWNLOAD,
+        /** REDOWNLOAD means the peer has given us a high-enough-work chain,
+         * and now we're redownloading the headers we saw before and trying to
+         * accept them */
+        REDOWNLOAD,
+        /** We're done syncing with this peer and can discard any remaining state */
+        FINAL
+    };
+
+    /** Return the current state of our download */
+    State GetState() const { return m_download_state; }
+
+    /** Start headers sync (via this download-twice mechanism)
+     * chain_start: best known fork point that the peer's headers branch from
+     * initial_headers: first batch of headers to process (assumes the caller
+     *                  has already verified the headers connect)
+     * minimum_required_work: amount of chain work required to accept the chain
+     */
+    std::optional<CBlockLocator> StartInitialDownload(const CBlockIndex* chain_start, const
+            std::vector<CBlockHeader>& initial_headers, const arith_uint256&
+            minimum_required_work, CBlockLocator&& chain_start_locator);
+
+    struct ProcessingResult {
+        std::optional<CBlockLocator> locator{std::nullopt};
+        std::vector<CBlockHeader> headers_to_process;
+        bool success{false};
+    };
+
+    /** Process a batch of headers, once a sync via this mechanism has started
+     *
+     * headers: headers that were received over the network for processing.
+     *          Assumes the caller has already verified the headers connect.
+     * full_headers_message: true if the message was at max capacity,
+     *                       indicating more headers may be available
+     * ProcessingResult.headers_to_process: will be filled in with any headers that the caller
+     *                       can process and validate now (because these returned
+     *                       headers are on a chain with sufficient work)
+     * ProcessingResult.success: set to false if an error is detected and the sync is
+     *                       aborted; true otherwise.
+     * ProcessingResult.locator: if present, the next locator to send in a
+     *                       getheaders message
+     */
+    ProcessingResult ProcessNextHeaders(const std::vector<CBlockHeader>&
```

Comment:
I'm doing the rename you suggest, but yes pulling headers out of `HeadersSyncState` needs to happen in lockstep with processing, so I think rather than add another API call to extract them, it makes the most sense to just return them.